xapian-core/weight/weight.cc

```diff
@@ -134,4 +135,81 @@ Weight::unserialise(const string &) const
     throw Xapian::UnimplementedError("unserialise() not supported for this Xapian::Weight subclass");
 }
 
+const Weight *
+Weight::create(const string & s, const Registry & reg)
+{
+    const char *p = s.c_str();
+    std::string scheme = "";
+
+    while (*p != ' ') {
+	if (*p == '\0') break;
+	scheme += *p;
+	p++;
+    }
+
+    if (scheme == "bb2")
+	scheme = "Xapian::BB2Weight";
+    else if (scheme == "bm25")
+	scheme = "Xapian::BM25Weight";
+    else if (scheme == "bm25plus")
+	scheme = "Xapian::BM25PlusWeight";
+    else if (scheme == "bool")
+	scheme = "Xapian::BoolWeight";
+    else if (scheme == "coord")
+	scheme = "Xapian::CoordWeight";
+    else if (scheme == "dlh")
+	scheme = "Xapian::DLHWeight";
+    else if (scheme == "dph")
+	scheme = "Xapian::DPHWeight";
+    else if (scheme == "ifb2")
+	scheme = "Xapian::IfB2Weight";
+    else if (scheme == "ineb2")
+	scheme = "Xapian::IneB2Weight";
+    else if (scheme == "inl2")
+	scheme = "Xapian::InL2Weight";
+    else if (scheme == "pl2")
+	scheme = "Xapian::PL2Weight";
+    else if (scheme == "pl2plus")
+	scheme = "Xapian::PL2PlusWeight";
+    else if (scheme == "tfidf")
+	scheme = "Xapian::TfIdfWeight";
+    else if (scheme == "trad")
+	scheme = "Xapian::TradWeight";
+    else
+	throw "Unknown $opt{weighting} setting: " + scheme;
+
+    if (C_isspace(*p))
+	return reg.get_weighting_scheme(scheme)->create_from_parameters(p);
+    else
+	return reg.get_weighting_scheme(scheme);
```

Comment:
Looking at the code I realise there's a problem with this short-cutting in the case where there are no parameters - `create_from_parameters()` creates a new object, while this case returns a pointer to an existing object.  So the caller can't delete the returned object, but if it doesn't delete it then it'll sometimes leak.  I think we have to always call `create_from_parameters()` here.